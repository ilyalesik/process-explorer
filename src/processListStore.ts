import {createEvent, createStore} from "effector";
import {ProcessDescriptor} from "ps-list";

export const processListStore = createStore<Array<ProcessDescriptor>>([]);

export const updateListEvent = createEvent<Array<ProcessDescriptor>>("updateListEvent");

processListStore.on(updateListEvent, (_, newValue) => {
    return newValue;
});
