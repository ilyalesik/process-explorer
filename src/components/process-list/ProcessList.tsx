import * as React from "react";
import * as s from "./processList.css";
const cc = require("classcat");
import * as R from "ramda";
import { $parentMap, killProcessEvent } from "../../processListStore";
import { useList } from "effector-react";
import { PercentageIndicator } from "../percentage-indicator/PercentageIndicator";
import { ProcessItemTemplate } from "../process-item/ProcessItemTemplate";
import { CloseButton } from "../close-button/CloseButton";
import { FieldProvider } from "./FieldProvider";

export const useChildrenProcessItems = pid => {
	const store = React.useRef($parentMap.map(R.propOr([], pid)));

	return useList(store.current, (item: number) => <ProcessItem pid={item} />);
};

export const ProcessItem = (props: { pid: number }) => {
	const { pid } = props;
	const childList = useChildrenProcessItems(pid);

	const name = (
		<FieldProvider pid={pid} field="name">
			{name => <span className={cc([s.text, s["text--white"]])}>{name}</span>}
		</FieldProvider>
	);

	const cpu = (
		<FieldProvider pid={pid} field="cpu">
			{cpu => typeof cpu === "number" && <PercentageIndicator value={cpu} className={s.text} />}
		</FieldProvider>
	);

	const memory = (
		<FieldProvider pid={pid} field="memory">
			{memory => typeof memory === "number" && <PercentageIndicator value={memory} className={s.text} />}
		</FieldProvider>
	);

	const close = <CloseButton onClick={() => killProcessEvent(pid)} />;

	return (
		<ProcessItemTemplate name={name} cpu={cpu} memory={memory} close={close}>
			{childList}
		</ProcessItemTemplate>
	);
};

export const ProcessList = () => {
	const childList = useChildrenProcessItems(0);

	return <div class={s.container}>{childList}</div>;
};
