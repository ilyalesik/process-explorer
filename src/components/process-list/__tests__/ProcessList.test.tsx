import * as React from "react";
import * as renderer from "react-test-renderer";
import { ProcessList } from "../ProcessList";
import { updateListEvent } from "../../../processListStore";

describe("<ProcessList />", () => {
	it("ProcessList renders correctly", () => {
		updateListEvent([
			{
				pid: 1,
				ppid: 0,
				name: "root",
				cpu: 0,
				memory: 0.1
			},
			{
				pid: 2,
				ppid: 1,
				name: "child1",
				cpu: 0,
				memory: 0
			},
			{
				pid: 3,
				ppid: 1,
				name: "child2",
				cpu: 2.14,
				memory: 2.3
			},
			{
				pid: 4,
				ppid: 3,
				name: "child3",
				cpu: 1.3,
				memory: 0
			}
		]);
		const tree = renderer.create(<ProcessList />);
		expect(tree).toMatchSnapshot();
	});
});
