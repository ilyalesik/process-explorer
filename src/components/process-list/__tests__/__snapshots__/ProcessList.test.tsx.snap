// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProcessList /> ProcessList renders correctly 1`] = `
<div
  class="container"
>
  <div
    className="container"
  >
    <div
      className="column name"
    >
      <span
        className="text text--white"
      >
        root
      </span>
    </div>
    <div
      className="column"
    >
      <p
        class="label"
      >
        cpu
      </p>
      <p
        className="text"
        style={
          Object {
            "color": "rgb(167, 213, 3)",
          }
        }
      >
        0
        %
      </p>
    </div>
    <div
      className="column"
    >
      <p
        class="label"
      >
        memory
      </p>
      <p
        className="text"
        style={
          Object {
            "color": "rgb(88, 252, 42)",
          }
        }
      >
        0.1
        %
      </p>
    </div>
    <div
      className="closeContainer"
    >
      <button
        className="closeButton"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    class="children"
  >
    <div
      className="container"
    >
      <div
        className="column name"
      >
        <span
          className="text text--white"
        >
          child1
        </span>
      </div>
      <div
        className="column"
      >
        <p
          class="label"
        >
          cpu
        </p>
        <p
          className="text"
          style={
            Object {
              "color": "rgb(167, 213, 3)",
            }
          }
        >
          0
          %
        </p>
      </div>
      <div
        className="column"
      >
        <p
          class="label"
        >
          memory
        </p>
        <p
          className="text"
          style={
            Object {
              "color": "rgb(167, 213, 3)",
            }
          }
        >
          0
          %
        </p>
      </div>
      <div
        className="closeContainer"
      >
        <button
          className="closeButton"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      class="children"
    />
    <div
      className="container"
    >
      <div
        className="column name"
      >
        <span
          className="text text--white"
        >
          child2
        </span>
      </div>
      <div
        className="column"
      >
        <p
          class="label"
        >
          cpu
        </p>
        <p
          className="text"
          style={
            Object {
              "color": "rgb(59, 255, 68)",
            }
          }
        >
          2.14
          %
        </p>
      </div>
      <div
        className="column"
      >
        <p
          class="label"
        >
          memory
        </p>
        <p
          className="text"
          style={
            Object {
              "color": "rgb(0, 191, 191)",
            }
          }
        >
          2.3
          %
        </p>
      </div>
      <div
        className="closeContainer"
      >
        <button
          className="closeButton"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      class="children"
    >
      <div
        className="container"
      >
        <div
          className="column name"
        >
          <span
            className="text text--white"
          >
            child3
          </span>
        </div>
        <div
          className="column"
        >
          <p
            class="label"
          >
            cpu
          </p>
          <p
            className="text"
            style={
              Object {
                "color": "rgb(0, 191, 191)",
              }
            }
          >
            1.3
            %
          </p>
        </div>
        <div
          className="column"
        >
          <p
            class="label"
          >
            memory
          </p>
          <p
            className="text"
            style={
              Object {
                "color": "rgb(167, 213, 3)",
              }
            }
          >
            0
            %
          </p>
        </div>
        <div
          className="closeContainer"
        >
          <button
            className="closeButton"
            onClick={[Function]}
          />
        </div>
      </div>
      <div
        class="children"
      />
    </div>
  </div>
</div>
`;
